p4-16/psa/PSA.mdk

```diff
@@ -1748,3 +1748,357 @@ types, as the P4 program author wishes.
+| CE2E         | clone from egress to egress |         |          |
+|--------------|-------------|---------------|--------------------|
+| NTP          | normal packet to port   | device at other | determined by the |
+|              |                         | end of port     | other device      |
+|--------------|-------------|---------------|--------------------|
+| NTCPU        | normal packet to CPU port | CPU | determined by CPU |
+|--------------|-------------|---------------|--------------------|
+~
+
+For egress packets, the choice of egress port, or the port to the CPU,
+is made by the immediately previous processing step (ingress for NU,
+NM, or CI2E packets, egress for CE2E packets), and egress processing
+cannot change that choice to a different port.  It can change the
+packet to RECIRC instead, or drop the packet.
+
+A single packet received by a PSA system from a port can result in 0,
+1, or many packets going out, all under control of the PSA program.
+For example, a single packet received from a port could cause all of
+the following to occur, if the PSA program so directed it:
+
++ The original packet received as NFP from port 2.  Ingress processing
+  creates a CI2E clone destined for the CPU port (copy 1), and a
+  multicast NM packet to multicast group 18, which is configured in
+  the PacketReplicationEngine to have copies made to ports 5 (copy 2)
+  and 7 (copy 3).
++ Copy 1 performs egress processing, which does not create a CE2E
+  clone, and sends the packet on path NTCPU to the CPU port.
++ Copy 2 performs egress processing, which creates a CE2E clone
+  destined for port 8 (copy 4), and sends a NTP packet to port 5.
++ Copy 3 performs egress processing, which does not create a CE2E
+  clone, and does a RECIRC back to ingress instead of sending the
+  packet to port 7 (copy 5).
++ Copy 4 performs egress processing, which does not create a CE2E
+  clone, and sends a NTP packet to port 8.
++ Copy 5 performs ingress processing, which does not create a CI2E
+  clone, and sends a NU packet destined for port 1 (copy 6).
++ Copy 6 performs egress processing, which does not create a CE2E
+  clone, and drops the packet instead of sending it to port 1.
+
+This is simply an example of what is possible given an appropriately
+written PSA program.  There is no need to use all of the packet paths
+available.  The numbering of the packet copies above is only for
+purposes of distinctly identifying each one in the example.  A PSA
+implementation is free to perform the steps above in many possible
+orders.
+
+There is no mechanism defined in PSA to prevent a single received
+packet from creating packets that continue to recirculate, resubmit,
+or clone from egress to egress indefinitely.  This can be prevented by
+suitable testing of your PSA program, and/or creating in your PSA
+program a "recirculate limit" metadata field that is carried along
+with copies of a packet, similar to the IPv4 Time To Live header
+field.
```

Comment:
I believe we should allow for target implementations to also limit the number of resubmit/recirc.

Reply:
Something like this?

"A PSA implementation may drop resubmitted, recirculated, or egress-to-egress clone packets after an implementation-specific maximum number of times of the same original packet.  If so, it should maintain counters of packets dropped for these reasons, and preferably record some debug information about the first few packets dropped for these reasons (perhaps only 1)."

Reply:
I added the text of the previous comment.

Reply:
Thanks, sounds good.